internal/staging/primitives/inclusionproof.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2019-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package primitives
+
+import (
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+// log2FloorMasks defines the masks to use when quickly calculating
+// floor(log2(x)) in a constant log2(32) = 5 steps, where x is a uint32, using
+// shifts.  They are derived from (2^(2^x) - 1) * (2^(2^x)), for x in 4..0.
+var log2FloorMasks = []uint32{0xffff0000, 0xff00, 0xf0, 0xc, 0x2}
+
+// fastLog2Ceil calculates and returns ceil(log2(x)) in a constant 5 steps.
+func fastLog2Ceil(n uint32) uint8 {
+	// The final answer will be one more when the input is greater than one and
+	// not a power of two.
+	rv := uint8(0)
+	if n&(n-1) != 0 {
+		rv++
+	}
+	exponent := uint8(16)
+	for i := 0; i < 5; i++ {
+		if n&log2FloorMasks[i] != 0 {
+			rv += exponent
+			n >>= exponent
+		}
+		exponent >>= 1
+	}
+	return rv
+}
+
+// GenerateInclusionProof treats the provided slice of hashes as leaves of a
+// merkle tree and generates and returns a merkle tree inclusion proof for the
+// given leaf index.  The proof can be used to efficiently prove the leaf
+// associated with a given leaf index is a member of the tree.
+//
+// A merkle tree inclusion proof consists of the ceil(log2(x)) intermediate
+// sibling hashes along the path from the target leaf to prove through the root
+// node.  The sibling hashes, along with the original leaf hash (and its
+// original leaf index), can be used to recalculate the merkle root which, in
+// turn, can be verified against a known good merkle root in order to prove the
+// leaf is actually a member of the tree at that position.
+//
+// For example, consider the following merkle tree:
+//
+//	         root = h1234 = h(h12 || h34)
+//	        /                            \
+//	  h12 = h(h1 || h2)             h34 = h(h3 || h4)
+//	   /             \               /             \
+//	  h1             h2             h3             h4
+//
+// Further, consider the goal is to prove inclusion of h3 at the 0-based leaf
+// index of 2.  The proof will consist of the sibling hashes h4 and h12.  On the
+// other hand, if the goal were to prove inclusion of h2 at the 0-based leaf
+// index of 1, the proof would consist of the sibling hashes h1 and h34.
+//
+// Specifying a leaf index that is out of range will return nil.
+func GenerateInclusionProof(leaves []chainhash.Hash, leafIndex uint32) []chainhash.Hash {
+	// Note that the proof is empty for both the zero and one leaf cases because
+	// there are no sibling hashes in either case.
+	if len(leaves) < 2 {
+		return nil
+	}
+
+	if leafIndex >= uint32(len(leaves)) {
+		return nil
```

Comment:
It looks like this case that was added afterward is not hit by the tests.

Reply:
Good call regarding the coverage.

For some additional details, this case wasn't added, rather the first if statement was updated to short circuit the result when there are less than 2 leaves since it will result in a `nil` regardless.  The test case that was hitting this branch was only using a single leaf, so it ended up being covered by the first if instead.

I've updated the test case to hit this branch again as well.